Trustroots/trustroots
# Convert translations from json to po and vice versa

#### Rationale

There are [tools](https://poedit.net/) [available](https://www.transifex.com/) for translating the gettext (.po) translations. Not so many for [i18next](https://locize.com/) format.

Therefore, let's keep both the `json` and `po` versions of the translations. This PR will help keep them in sync. Developers can generate `json` files and translators can translate `po` files.

#### Proposed Changes

* a script to convert translation files in `public/locales/{locale}/` from json to po and vice versa
* pre-commit hook to do the conversion, when translation files are changed.

There may be more elegant solutions than this. I'm interested to hear about them, please!

_If you can think of a better name or location for the tests, please share._

#### TODO

- [x] write the conversion scripts
- [x] make a pre-commit hook which will automatically convert json <-> po and keep them in sync
- [x] refactor, comment
- [x] format the output json file (has 4 spaces for tab instead of 2)

#### Testing Instructions

0. `git switch convert-translations-po`
0. json to po
    - create or edit files in `public/locales/{locale}/{translation}.json`
    - `git add . && git commit -m "test"`
    - see that their `.po` counterparts were generated/updated
0. po to json
    - create or edit files in `public/locales/{locale}/{translation}.po`
    - `git add . && git commit -m "test"`
    - see that their `.json` counterparts were generated/updated
0. both directions
    - create or edit various `json` and `po` files in `public/locales`
    - `git add . && git commit -m "test"`
    - see that their counterparts were generated/updated
0. both directions with conflict
    - create or edit both `json` and `po` with the same name and locale, e.g. `public/locales/en/volunteering.{json|po}`
    - `git add . && git commit -m "test"`
    - see that the pre-hook will complain about conflict between json and po file
0. use script directly
    - run `npm run convert-translations public/locales/**/*.json`
    - see that everything got converted to po.
